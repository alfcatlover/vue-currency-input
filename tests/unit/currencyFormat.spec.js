import CurrencyFormat from '../../src/currencyFormat'

describe('CurrencyFormat', () => {
  describe('i18n', () => {
    it('de-DE_EUR', () => {
      expect(new CurrencyFormat({ locale: 'de-DE', currency: 'EUR' })).toMatchSnapshot()
    })

    it('es-ES_EUR', () => {
      expect(new CurrencyFormat({ locale: 'es-ES', currency: 'EUR' })).toMatchSnapshot()
    })

    it('nl-NL_EUR', () => {
      expect(new CurrencyFormat({ locale: 'nl-NL', currency: 'EUR' })).toMatchSnapshot()
    })

    it('en-US_USD', () => {
      expect(new CurrencyFormat({ locale: 'en-US', currency: 'USD' })).toMatchSnapshot()
    })

    it('zh_CNY', () => {
      expect(new CurrencyFormat({ locale: 'zh', currency: 'CNY' })).toMatchSnapshot()
    })

    it('en-GB_GBP', () => {
      expect(new CurrencyFormat({ locale: 'en-GB', currency: 'GBP' })).toMatchSnapshot()
    })

    it('en-IN_INR', () => {
      expect(new CurrencyFormat({ locale: 'en-GB', currency: 'INR' })).toMatchSnapshot()
    })

    it('pt_BRL', () => {
      expect(new CurrencyFormat({ locale: 'pt', currency: 'BRL' })).toMatchSnapshot()
    })

    it('ja_JPY', () => {
      expect(new CurrencyFormat({ locale: 'ja', currency: 'JPY' })).toMatchSnapshot()
    })
  })

  it('custom prefix/suffix', () => {
    expect(new CurrencyFormat({ locale: 'de-DE', currency: { prefix: '₿ ' } })).toMatchSnapshot()
    expect(new CurrencyFormat({ locale: 'de-DE', currency: { suffix: ' Euro' } })).toMatchSnapshot()
    expect(new CurrencyFormat({ locale: 'de-DE', currency: null })).toMatchSnapshot()
  })

  it('custom precision', () => {
    expect(new CurrencyFormat({ locale: 'de-DE', currency: 'EUR', precision: 0 })).toMatchSnapshot()
    expect(new CurrencyFormat({ locale: 'de-DE', currency: 'EUR', precision: { min: 0, max: 0 } })).toMatchSnapshot()
    expect(new CurrencyFormat({ locale: 'de-DE', currency: null, precision: { min: 0, max: 2 } })).toMatchSnapshot()
    expect(new CurrencyFormat({ locale: 'de-DE', currency: { suffix: ' €' }, precision: { min: 0, max: 2 } })).toMatchSnapshot()
    expect(new CurrencyFormat({ locale: 'ja', currency: 'JPY', precision: { min: 2, max: 2 } })).toMatchSnapshot()
  })

  describe('parse', () => {
    it('returns null if the value is empty', () => {
      expect(new CurrencyFormat({}).parse('')).toBeNull()
      expect(new CurrencyFormat({}).parse(' ')).toBeNull()
    })

    it('returns null if the value is invalid', () => {
      expect(new CurrencyFormat({ locale: 'en' }).parse('-')).toBeNull()
      expect(new CurrencyFormat({ locale: 'en' }).parse('123e-1')).toBeNull()
      expect(new CurrencyFormat({ locale: 'en' }).parse('0x11')).toBeNull()
      expect(new CurrencyFormat({ locale: 'en' }).parse('0b11')).toBeNull()
      expect(new CurrencyFormat({ locale: 'en' }).parse('0o11')).toBeNull()
      expect(new CurrencyFormat({ locale: 'en' }).parse('1.2e1')).toBeNull()
      expect(new CurrencyFormat({ locale: 'en' }).parse('1.23.4')).toBeNull()
    })

    it('returns the parsed number if the value conforms to the currency format config', () => {
      expect(new CurrencyFormat({ locale: 'en' }).parse('1234')).toBe(1234)
      expect(new CurrencyFormat({ locale: 'en', precision: 3 }).parse('1234', true)).toBe(1234000)
      expect(new CurrencyFormat({ locale: 'en' }).parse('1,234,567')).toBe(1234567)
      expect(new CurrencyFormat({ locale: 'en' }).parse('-1,234,567')).toBe(-1234567)
      expect(new CurrencyFormat({ locale: 'en' }).parse('−1,234,567')).toBe(-1234567)
      expect(new CurrencyFormat({ locale: 'de' }).parse('−1234567,89')).toBe(-1234567.89)
      expect(new CurrencyFormat({ locale: 'en', currency: 'USD' }).parse('$1,234,567')).toBe(1234567)
      expect(new CurrencyFormat({ locale: 'de', currency: 'EUR' }).parse('1234 €')).toBe(1234)
      expect(new CurrencyFormat({ locale: 'en' }).parse('-1234')).toBe(-1234)
      expect(new CurrencyFormat({ locale: 'en', currency: 'USD' }).parse('-$1234')).toBe(-1234)
      expect(new CurrencyFormat({ locale: 'de', currency: 'EUR' }).parse('-1234 €')).toBe(-1234)
      expect(new CurrencyFormat({ locale: 'ja', currency: 'JPY' }).parse('￥123,456')).toBe(123456)
      expect(new CurrencyFormat({ locale: 'en' }).parse('.5')).toBe(null)
      expect(new CurrencyFormat({ locale: 'en' }).parse('0.5')).toBe(0.5)
      expect(new CurrencyFormat({ locale: 'en' }).parse('1234.50')).toBe(1234.5)
      expect(new CurrencyFormat({ locale: 'en' }).parse('1234.00')).toBe(1234)
      expect(new CurrencyFormat({ locale: 'en', currency: 'USD' }).parse('$1,234.50')).toBe(1234.5)
      expect(new CurrencyFormat({ locale: 'en', currency: 'USD' }).parse('$1,234.50', true)).toBe(123450)
      expect(new CurrencyFormat({ locale: 'de', currency: 'EUR' }).parse('1.234,50 €')).toBe(1234.5)
    })

    it('returns null if the value does not conform to the currency format config', () => {
      expect(new CurrencyFormat({ locale: 'en' }).parse('1234,5')).toBeNull()
      expect(new CurrencyFormat({ locale: 'de' }).parse('1,234,567.89')).toBeNull()
      expect(new CurrencyFormat({ locale: 'de', currency: 'EUR' }).parse('$1234')).toBeNull()
      expect(new CurrencyFormat({ locale: 'en', currency: 'USD' }).parse('1234 €')).toBeNull()
    })
  })
})
