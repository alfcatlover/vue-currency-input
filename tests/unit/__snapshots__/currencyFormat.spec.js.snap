// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyFormat custom precision 1`] = `
CurrencyFormat {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 0,
  "minimumFractionDigits": 0,
  "minusSymbol": "-",
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " €",
}
`;

exports[`CurrencyFormat custom precision 2`] = `
CurrencyFormat {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 0,
  "minimumFractionDigits": 0,
  "minusSymbol": "-",
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " €",
}
`;

exports[`CurrencyFormat custom precision 3`] = `
CurrencyFormat {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 0,
  "minusSymbol": "-",
  "negativePrefix": "-",
  "prefix": "",
  "suffix": "",
}
`;

exports[`CurrencyFormat custom precision 4`] = `
CurrencyFormat {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 0,
  "minusSymbol": "-",
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " €",
}
`;

exports[`CurrencyFormat custom precision 5`] = `
CurrencyFormat {
  "decimalSymbol": undefined,
  "groupingSymbol": ",",
  "maximumFractionDigits": 0,
  "minimumFractionDigits": 0,
  "minusSymbol": "-",
  "negativePrefix": "-￥",
  "prefix": "￥",
  "suffix": "",
}
`;

exports[`CurrencyFormat custom prefix/suffix 1`] = `
CurrencyFormat {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-₿ ",
  "prefix": "₿ ",
  "suffix": "",
}
`;

exports[`CurrencyFormat custom prefix/suffix 2`] = `
CurrencyFormat {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " Euro",
}
`;

exports[`CurrencyFormat custom prefix/suffix 3`] = `
CurrencyFormat {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-",
  "prefix": "",
  "suffix": "",
}
`;

exports[`CurrencyFormat i18n de-DE_EUR 1`] = `
CurrencyFormat {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " €",
}
`;

exports[`CurrencyFormat i18n en-GB_GBP 1`] = `
CurrencyFormat {
  "decimalSymbol": ".",
  "groupingSymbol": ",",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-£",
  "prefix": "£",
  "suffix": "",
}
`;

exports[`CurrencyFormat i18n en-IN_INR 1`] = `
CurrencyFormat {
  "decimalSymbol": ".",
  "groupingSymbol": ",",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-₹",
  "prefix": "₹",
  "suffix": "",
}
`;

exports[`CurrencyFormat i18n en-US_USD 1`] = `
CurrencyFormat {
  "decimalSymbol": ".",
  "groupingSymbol": ",",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-$",
  "prefix": "$",
  "suffix": "",
}
`;

exports[`CurrencyFormat i18n es-ES_EUR 1`] = `
CurrencyFormat {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " €",
}
`;

exports[`CurrencyFormat i18n ja_JPY 1`] = `
CurrencyFormat {
  "decimalSymbol": undefined,
  "groupingSymbol": ",",
  "maximumFractionDigits": 0,
  "minimumFractionDigits": 0,
  "minusSymbol": "-",
  "negativePrefix": "-￥",
  "prefix": "￥",
  "suffix": "",
}
`;

exports[`CurrencyFormat i18n nl-NL_EUR 1`] = `
CurrencyFormat {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "€ -",
  "prefix": "€ ",
  "suffix": "",
}
`;

exports[`CurrencyFormat i18n pt_BRL 1`] = `
CurrencyFormat {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-R$ ",
  "prefix": "R$ ",
  "suffix": "",
}
`;

exports[`CurrencyFormat i18n zh_CNY 1`] = `
CurrencyFormat {
  "decimalSymbol": ".",
  "groupingSymbol": ",",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-¥",
  "prefix": "¥",
  "suffix": "",
}
`;
